File: projects/Vibe_XASR/windows_build/src/VibeXASR.Windows/Lexicon/Defiller.cs
using System.Text.RegularExpressions;

namespace VibeXASR.Windows.Lexicon;

/// <summary>
/// Remove Chinese filler words from FINAL dictation — the touch that makes speech read like
/// writing. Conservative: deletes pure interjections
/// (嗯/呃/唉…) and collapses ≥4× repeats, so genuine reduplications (看看 / 想想 / 好好) and short
/// counting (三三三) stay, and a single meaningful 那个 / 就是 is left intact — only stutter-style
/// repeats fold. NOTE: 額/额/诶 are NOT interjections — they occur in real words (金额 / 额外 / 余额).
/// Faithful port of macOS Defiller.swift.
/// </summary>
internal static class Defiller
{
    // 額/额/诶 are deliberately NOT here — they occur in real words (金额 / 额外 / 余额).
    private const string Interjections = "嗯呃唉欸喔噢";
    private static readonly string[] RepeatWords = { "那个", "这个", "就是", "然后" };

    public static string Clean(string? text)
    {
        if (string.IsNullOrEmpty(text)) return text ?? string.Empty;
        var s = text;
        // 1) pure interjections (and any run of them)
        s = Regex.Replace(s, "[" + Interjections + "]+", "");
        // 2) collapse a character repeated ≥4× → once (≤3× e.g. 看看/三三三 untouched; only true stutters like 这这这这 fold)
        s = Regex.Replace(s, @"(.)\1{3,}", "$1");
        // 3) collapse stutter repeats of common fillers (≥2×) → once
        foreach (var w in RepeatWords) s = Regex.Replace(s, "(?:" + w + "){2,}", w);
        // 4) tidy punctuation left behind by removed interjections
        s = Regex.Replace(s, @"^[，,、。!?！？\s]+", "");
        s = Regex.Replace(s, "([，,])[，,]+", "$1");
        return s;
    }
}

File: projects/Vibe_XASR/macos_build/native/app/Sources/VibeIME/Defiller.swift
import Foundation

/// Remove Chinese filler words from FINAL dictation — the touch that makes speech
/// read like writing. Conservative:
/// deletes pure interjections (嗯/呃/唉…) and collapses ≥4× repeats, so genuine
/// reduplications (看看 / 想想 / 好好) and short counting (三三三) stay, and a single
/// meaningful 那个 / 就是 is left intact — only stutter-style repeats are folded.
/// NOTE: 額/额/诶 are deliberately NOT interjections — they occur in real words
/// (金额 / 额外 / 余额), so stripping them mangled those.
enum Defiller {
    private static let interjections = "嗯呃唉欸喔噢"
    private static let repeatWords = ["那个", "这个", "就是", "然后"]

    static func clean(_ text: String) -> String {
        var s = text
        // 1) pure interjections (and any run of them)
        s = s.replacingOccurrences(of: "[\(interjections)]+", with: "", options: .regularExpression)
        // 2) collapse a character repeated ≥4× → once (≤3× e.g. 看看/三三三 untouched;
        //    only true stutters like 这这这这 fold)
        s = s.replacingOccurrences(of: "(.)\\1{3,}", with: "$1", options: .regularExpression)
        // 3) collapse stutter repeats of common fillers (≥2×) → once
        for w in repeatWords {
            s = s.replacingOccurrences(of: "(?:\(w)){2,}", with: w, options: .regularExpression)
        }
        // 4) tidy punctuation left behind by removed interjections
        s = s.replacingOccurrences(of: "^[，,、。!?！？\\s]+", with: "", options: .regularExpression)
        s = s.replacingOccurrences(of: "([，,])[，,]+", with: "$1", options: .regularExpression)
        return s
    }
}
